lib/spack/spack/ci.py

```diff
@@ -546,6 +486,235 @@ def get_spec_filter_list(env, affected_pkgs):
+
+        for match_attrs in section["submapping"]:
+            attrs = cfg.InternalConfigScope._process_dict_keyname_overrides(match_attrs)
+            for match_string in match_attrs["match"]:
+                if _spec_matches(spec, match_string):
+                    matched = True
+                    if "build-job-remove" in match_attrs:
+                        spack.config.remove_yaml(dest, attrs["build-job-remove"])
+                    if "build-job" in match_attrs:
+                        spack.config.merge_yaml(dest, attrs["build-job"])
+                    break
+            if matched and only_first:
+                break
+
+        return dest
+
+    # Generate IR from the configs
+    def generate_ir(self):
+        """Generate the IR from the Spack CI configurations."""
+
+        jobs = self.ir["jobs"]
+
+        # Implicit job defaults
+        defaults = [
+            {
+                "build-job": {
+                    "script": [
+                        "cd {env_dir}",
+                        "spack env activate --without-view .",
+                        "spack ci rebuild",
+                    ],
+                }
+            },
+        ]
+
+        # Job overrides
+        overrides = [
+            # Configurable reindex script
+            {
+                "reindex-job": {
+                    "script": [
+                        "spack buildcache update-index --keys --mirror-url {index_target_mirror}",
+                    ],
+                }
+            },
+            {
+                "reindex-job": {
+                    "script": "",
+                },
+            },
+            # Configurable cleanup script
+            {
+                "cleanup-job": {
+                    "script": [
+                        "spack -d mirror destroy --mirror-url {mirror_prefix}/$CI_PIPELINE_ID",
+                    ]
+                }
+            },
+            # Add signing job tags
+            {
+                "signing-job": {
+                    "tags": ["aws", "protected", "notary"],
+                }
+            },
+            # Remove reserved tags
+            {
+                "any-job-remove": {
+                    "tags": SPACK_RESERVED_TAGS,
+                }
+            },
+        ]
+
+        pipeline_gen = overrides + self.ci_config.get("pipeline-gen", []) + defaults
+
+        for section in reversed(pipeline_gen):
+            name = self.__is_named(section)
+            has_submapping = "submapping" in section
+            section = cfg.InternalConfigScope._process_dict_keyname_overrides(section)
+
+            if name:
+                remove_job_name = self.__job_name(name, suffix="-remove")
+                merge_job_name = self.__job_name(name)
+                do_remove = remove_job_name in section
+                do_merge = merge_job_name in section
+
+                def _apply_section(attrs, section):
```

Comment:
Why define this function within the loop?  Could you move it outside to be defined only once if you pass in `do_merge` and `do_remove` as parameters?

Reply:
I am using it more like a lambda with a capture list, I thought it was a little more readable to just say "apply the section to the job attributes" without the extra args. If this isn't idiomatic python I can change it though, I didn't think it cost that much to define a function in a loop.

Reply:
I thought it was also weird to shadow the closure variable `section` with the function parameter of the same name, but whatever.  I'm not an idiomatic python person, I just thought this was weird and suspicious when I was debugging the extra `:` getting written out in the final gitlab yaml.  But it turned out to have nothing to do with the issue, so I guess it's fine the way it is.

Reply:
ah, yeah, shadowing is not really good...I will at least change the name here to make it clear it is the passed parameter not the loop parameter.